### PR TITLE
Enable i915kms to work on 5.5-wip

### PR DESCRIPTION
**Pull request by neelchauhan**
This branch enables i915kms to work on the 5.5-wip branch.

Tested on:

 * HP Spectre x360 13-ap0053dx (WhiskeyLake)
 * Blank screen ~~/freeze~~ : HP Spectre x360 13t-aw200 (TigerLake) - ~~mirrors 5.4 behavior~~ (**UPDATE**: Does not freeze anymore, but still gives blank screen)

This is my first drm-kmod patch, so sorry if it's not "optimal".

### CODE REVIEW

File: drivers/gpu/drm/i915/i915_drv.c [hunk 1]
```diff
@@ -2770,10 +2770,10 @@ static const struct drm_ioctl_desc i915_ioctls[] = {
...
 	DRM_IOCTL_DEF_DRV(I915_GEM_CONTEXT_GETPARAM, i915_gem_context_getparam_ioctl, DRM_RENDER_ALLOW),
 	DRM_IOCTL_DEF_DRV(I915_GEM_CONTEXT_SETPARAM, i915_gem_context_setparam_ioctl, DRM_RENDER_ALLOW),
-	DRM_IOCTL_DEF_DRV(I915_PERF_OPEN, i915_perf_open_ioctl, DRM_RENDER_ALLOW),
-	DRM_IOCTL_DEF_DRV(I915_PERF_ADD_CONFIG, i915_perf_add_config_ioctl, DRM_RENDER_ALLOW),
-	DRM_IOCTL_DEF_DRV(I915_PERF_REMOVE_CONFIG, i915_perf_remove_config_ioctl, DRM_RENDER_ALLOW),
-	DRM_IOCTL_DEF_DRV(I915_QUERY, i915_query_ioctl, DRM_RENDER_ALLOW),
+	//DRM_IOCTL_DEF_DRV(I915_PERF_OPEN, i915_perf_open_ioctl, DRM_RENDER_ALLOW),
+	//DRM_IOCTL_DEF_DRV(I915_PERF_ADD_CONFIG, i915_perf_add_config_ioctl, DRM_RENDER_ALLOW),
+	//DRM_IOCTL_DEF_DRV(I915_PERF_REMOVE_CONFIG, i915_perf_remove_config_ioctl, DRM_RENDER_ALLOW),
+	//DRM_IOCTL_DEF_DRV(I915_QUERY, i915_query_ioctl, DRM_RENDER_ALLOW),
```
Comment (unrelentingtech) [inline drivers/gpu/drm/i915/i915_drv.c:2776]: `#ifdef __linux__` these instead of commenting. What's the error btw?

**Review by evadot — COMMENTED** [review]
I'm not against refactoring the Makefile but this needs to be done in a separate commit with a proper commit message.

File: drivers/gpu/drm/i915/i915_drv.c (hunk 1 above)
Comment (evadot) [reply drivers/gpu/drm/i915/i915_drv.c:2776]: Same question

**Comment by evadot** [comment]
There is still some stuff needed in linuxkpi for 5.5 on i915, just grep for lkpi_todo

File: i915/Makefile [hunk 2]
```diff
@@ -12,11 +12,11 @@ SRCS=	\
 	i915_buddy.c \
 	i915_cmd_parser.c \
 	i915_drv.c \
-	i915_getparam.c \
```
Comment (evadot) [inline i915/Makefile:15]: Does this refactoring adds any file ?
It's hard to see ...
Also some of the new ordering doesn't match what I have with emacs M-x sort-lines which is usually what I use, so I don't like that.
If any files are missing, please add them and just that.

**Review by evadot — CHANGES_REQUESTED** [review]

**Review by evadot — CHANGES_REQUESTED** [review]
Why ?
Please add the reason in the commit message.

File: i915/Makefile (hunk 2 above)
Comment (neelchauhan) [reply i915/Makefile]: It does add a few files. I have reverted to the old Makefile and removed the respective commits (sorry for force pushing), and only added the needed files.

**Comment by neelchauhan** [comment]
The work so far:

Intel TigerLake:

    drmn0: <drmn> on vgapci0
    [drm] i915.alpha_support is deprecated, use i915.force_probe=9a49 instead
    vgapci0: child drmn0 requested pci_enable_io
    vgapci0: child drmn0 requested pci_enable_io
    [drm] Unable to create a private tmpfs mount, hugepage support will be disabled(-19).
    Failed to add WC MTRR for [0x4000000000-0x400fffffff]: -22; performance may suffer
    [drm] Got stolen memory base 0x0, size 0x0
    [drm] Supports vblank timestamp caching Rev 2 (21.10.2013).
    [drm] Driver supports precise vblank timestamp query.
    drmn0: could not load firmware image 'i915/tgl_dmc_ver2_04.bin'
    drmn0: Failed to load DMC firmware i915/tgl_dmc_ver2_04.bin. Disabling runtime power management.
    drmn0: DMC firmware homepage: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/i915<6>[drm] Connector eDP-1: get mode from tunables:
    [drm]   - kern.vt.fb.modes.eDP-1
    [drm]   - kern.vt.fb.default_mode
    sysctl_warn_reuse: can't re-use a leaf (hw.dri.debug)!
    [drm] Initialized i915 1.6.0 20190822 for drmn0 on minor 0
    WARNING: Device "fb" is Giant locked and may be deleted before FreeBSD 14.0.
    VT: Replacing driver "efifb" with new "fb".
    start FB_INFO:
    type=11 height=1080 width=1920 depth=32
    cmsize=16 size=8294400
    pbase=0x4000040000 vbase=0xfffffe011ac40000
    name=drmn0 flags=0x0 stride=7680 bpp=32
    cmap[0]=0 cmap[1]=7f0000 cmap[2]=7f00 cmap[3]=c4a000
    end FB_INFO
    drmn0: fb0: i915drmfb frame buffer device
    Unclaimed read from register 0x64040

Now the computer does not freeze. It still gives a blank screen, but I can use the caps lock key and power button once (as opposed to pressing) to ACPI shutdown.

On WhiskeyLake:

    drmn0: <drmn> on vgapci0
    vgapci0: child drmn0 requested pci_enable_io
    vgapci0: child drmn0 requested pci_enable_io
    [drm] Unable to create a private tmpfs mount, hugepage support will be disabled(-19).
    Failed to add WC MTRR for [0x4000000000-0x400fffffff]: -22; performance may suffer
    [drm] Got stolen memory base 0x59000000, size 0x4000000
    [drm] Supports vblank timestamp caching Rev 2 (21.10.2013).
    [drm] Driver supports precise vblank timestamp query.
    [drm] Connector eDP-1: get mode from tunables:
    [drm]   - kern.vt.fb.modes.eDP-1
    [drm]   - kern.vt.fb.default_mode
    drmn0: successfully loaded firmware image 'i915/kbl_dmc_ver1_04.bin'
    [drm] Finished loading DMC firmware i915/kbl_dmc_ver1_04.bin (v1.4)
    [drm] Connector DP-1: get mode from tunables:
    [drm]   - kern.vt.fb.modes.DP-1
    [drm]   - kern.vt.fb.default_mode
    [drm] Connector DP-2: get mode from tunables:
    [drm]   - kern.vt.fb.modes.DP-2
    [drm]   - kern.vt.fb.default_mode
    sysctl_warn_reuse: can't re-use a leaf (hw.dri.debug)!
    [drm] Initialized i915 1.6.0 20190822 for drmn0 on minor 0
    WARNING: Device "fb" is Giant locked and may be deleted before FreeBSD 14.0.
    VT: Replacing driver "efifb" with new "fb".
    taskqueue_drain with the following non-sleepable locks held:
    exclusive sleep mutex vtdev (vtdev) r = 0 (0xffffffff81ae8d68) locked @ /usr/src/sys/dev/vt/vt_core.c:3012
    stack backtrace:
    #0 0xffffffff80c803e1 at witness_debugger+0x71
    #1 0xffffffff80c814fd at witness_warn+0x40d
    #2 0xffffffff80c72e19 at taskqueue_drain+0x39
    #3 0xffffffff834d8693 at vt_kms_postswitch+0x73
    #4 0xffffffff80a6440d at vt_fb_init+0xfd
    #5 0xffffffff80a6bb18 at vt_replace_backend+0x118
    #6 0xffffffff80a64513 at vt_fb_attach+0x13
    #7 0xffffffff834d9168 at linux_register_framebuffer+0x5e8
    #8 0xffffffff834e0a19 at __drm_fb_helper_initial_config_and_unlock+0x419
    #9 0xffffffff8317f400 at intel_fbdev_initial_config_async+0x20
    #10 0xffffffff830750f4 at i915_driver_probe+0x1114
    #11 0xffffffff8308facd at i915_pci_probe+0x5d
    #12 0xffffffff80e6caf6 at linux_pci_attach_device+0x2d6
    #13 0xffffffff80c4b87a at device_attach+0x3ca
    #14 0xffffffff80c4b420 at device_probe_and_attach+0x70
    #15 0xffffffff80c4d2f8 at bus_generic_driver_added+0x58
    #16 0xffffffff80c49079 at devclass_driver_added+0x39
    #17 0xffffffff80c48fb7 at devclass_add_driver+0x147
    Sleeping on "tq_drain" with the following non-sleepable locks held:
    exclusive sleep mutex vtdev (vtdev) r = 0 (0xffffffff81ae8d68) locked @ /usr/src/sys/dev/vt/vt_core.c:3012
    stack backtrace:
    #0 0xffffffff80c803e1 at witness_debugger+0x71
    #1 0xffffffff80c814fd at witness_warn+0x40d
    #2 0xffffffff80c1cb8b at _sleep+0x5b
    #3 0xffffffff80c72edb at taskqueue_drain+0xfb
    #4 0xffffffff834d8693 at vt_kms_postswitch+0x73
    #5 0xffffffff80a6440d at vt_fb_init+0xfd
    #6 0xffffffff80a6bb18 at vt_replace_backend+0x118
    #7 0xffffffff80a64513 at vt_fb_attach+0x13
    #8 0xffffffff834d9168 at linux_register_framebuffer+0x5e8
    #9 0xffffffff834e0a19 at __drm_fb_helper_initial_config_and_unlock+0x419
    #10 0xffffffff8317f400 at intel_fbdev_initial_config_async+0x20
    #11 0xffffffff830750f4 at i915_driver_probe+0x1114
    #12 0xffffffff8308facd at i915_pci_probe+0x5d
    #13 0xffffffff80e6caf6 at linux_pci_attach_device+0x2d6
    #14 0xffffffff80c4b87a at device_attach+0x3ca
    #15 0xffffffff80c4b420 at device_probe_and_attach+0x70
    #16 0xffffffff80c4d2f8 at bus_generic_driver_added+0x58
    #17 0xffffffff80c49079 at devclass_driver_added+0x39
    lock order reversal: (Giant after non-sleepable)
     1st 0xffffffff81ae8d68 vtdev (vtdev, sleep mutex) @ /usr/src/sys/dev/vt/vt_core.c:3012
     2nd 0xffffffff81a02800 Giant (Giant, sleep mutex) @ /usr/src/sys/kern/kern_synch.c:228
    lock order Giant -> vtdev established at:
    #0 0xffffffff80c7f7bd at witness_checkorder+0x46d
    #1 0xffffffff80bed054 at __mtx_lock_flags+0x94
    #2 0xffffffff80a6b103 at vt_upgrade+0x363
    #3 0xffffffff80b9c99c at mi_startup+0xec
    #4 0xffffffff8038002c at btext+0x2c
    lock order vtdev -> Giant attempted at:
    #0 0xffffffff80c8011c at witness_checkorder+0xdcc
    #1 0xffffffff80bed054 at __mtx_lock_flags+0x94
    #2 0xffffffff80c1ce7f at _sleep+0x34f
    #3 0xffffffff80c72edb at taskqueue_drain+0xfb
    #4 0xffffffff834d8693 at vt_kms_postswitch+0x73
    #5 0xffffffff80a6440d at vt_fb_init+0xfd
    #6 0xffffffff80a6bb18 at vt_replace_backend+0x118
    #7 0xffffffff80a64513 at vt_fb_attach+0x13
    #8 0xffffffff834d9168 at linux_register_framebuffer+0x5e8
    #9 0xffffffff834e0a19 at __drm_fb_helper_initial_config_and_unlock+0x419
    #10 0xffffffff8317f400 at intel_fbdev_initial_config_async+0x20
    #11 0xffffffff830750f4 at i915_driver_probe+0x1114
    #12 0xffffffff8308facd at i915_pci_probe+0x5d
    #13 0xffffffff80e6caf6 at linux_pci_attach_device+0x2d6
    #14 0xffffffff80c4b87a at device_attach+0x3ca
    #15 0xffffffff80c4b420 at device_probe_and_attach+0x70
    #16 0xffffffff80c4d2f8 at bus_generic_driver_added+0x58
    #17 0xffffffff80c49079 at devclass_driver_added+0x39
    start FB_INFO:
    type=11 height=2160 width=3840 depth=32
    cmsize=16 size=33177600
    pbase=0x4000340000 vbase=0xfffffe0155540000
    name=drmn0 flags=0x0 stride=15360 bpp=32
    cmap[0]=0 cmap[1]=7f0000 cmap[2]=7f00 cmap[3]=c4a000
    end FB_INFO
    drmn0: fb0: i915drmfb frame buffer device

**Comment by unrelentingtech** [comment]
> except for wait_bit.h 

`wait_var_event_interruptible` can probably be defined [similarly to `wait_var_event_killable`](https://github.com/freebsd/freebsd-src/blob/4281bfec36285e2212f41568459c077bf4dbd91c/sys/compat/linuxkpi/common/include/linux/wait_bit.h#L40-L41) (though hmm what's up with ignoring the var and using one queue for everything?)

`clear_and_wake_up_bit` is also already defined.

**Comment by neelchauhan** [comment]
I believe this PR can be reviewed with it's updates. Sorry for the force pushes.

It's not surprising that TigerLake isn't supported, so TL shouldn't be a goal here.

**Comment by unrelentingtech** [comment]
Force pushes are the right way to work on a PR, nothing to be sorry for :D

Tiger Lake is only declared stable and enabled out of the box in 5.7, yeah.

---

Also, the LKPI_TODOs that aren't actually based on GPL Linux code should be submitted to base…

**Comment by neelchauhan** [comment]
> Force pushes are the right way to work on a PR, nothing to be sorry for :D

Thanks!

In the years prior, I was a heavy Tor contributor, and the Tor people weren't exactly too happy with force pushes in dev PRs, so that's why I'm "sorry". I'm glad we are more happy with force pushes in PRs.

> Tiger Lake is only declared stable and enabled out of the box in 5.7, yeah.

Darn, but not surprising. Don't worry, I'll still work on this PR and contribute and can test on WhiskeyLake. But I hope when 5.5 is done we skip to at least 5.7 if not the latest Linux (LTS or non-LTS), so then we can easily support TigerLake.

> Also, the LKPI_TODOs that aren't actually based on GPL Linux code should be submitted to base…

I'll do this. No exact ETA, but I'll attempt to create Phabricator reviews manu@ can review.

**Comment by neelchauhan** [comment]
For the linuxkpi updates that aren't GPL'd:

 * https://reviews.freebsd.org/D29987 (`wait_event_interruptible()` macro)

 * https://reviews.freebsd.org/D29988 (`atomic_dec_and_lock_irqsave()`)

**Comment by neelchauhan** [comment]
The above reviews have been committed to src as `ce65353ac1a17677bef03e96df8da967c9086743` and `057f145aae9a3528ccd722c8f566d740d5cafcbb`.

**Comment by unrelentingtech** [comment]
Nice. Are `tasklet_disable_nosync` and `rcu_replace_pointer` "GPL"? Simple 2/3-line definitions that are responsible for extra files here...

**Comment by neelchauhan** [comment]
> Nice. Are `tasklet_disable_nosync` and `rcu_replace_pointer` "GPL"? Simple 2/3-line definitions that are responsible for extra files here...

To my best understanding, yes:

 * https://elixir.bootlin.com/linux/v5.5.19/source/include/linux/interrupt.h#L664
 * https://elixir.bootlin.com/linux/v5.5.19/source/include/linux/rcupdate.h#L396

File: linuxkpi/gplv2/include/linux/rcupdate.h [hunk 3]
```diff
@@ -1,4 +1,13 @@
+#ifndef	__GPLV2_RCUPDATE_H__
+#define	__GPLV2_RCUPDATE_H__
+
 #include_next <linux/rcupdate.h>
 
-/* LKPI_TODO */
-#define	rcu_replace_pointer(rcu, ptr, c) 0
+#define	rcu_replace_pointer(rcu, ptr, c)				\
```
Comment (evadot) [inline linuxkpi/gplv2/include/linux/rcupdate.h:6]: This should be submitted for linuxkpi in base.

File: linuxkpi/gplv2/include/linux/interrupt.h [hunk 4]
```diff
@@ -1,4 +1,12 @@
+#ifndef	__GPLV2_INTERRUPT_H__
+#define	__GPLV2_INTERRUPT_H__
+
 #include_next <linux/interrupt.h>
 
-/* LKPI_TODO */
-#define	tasklet_disable_nosync(x) 0
+static inline void tasklet_disable_nosync(struct tasklet_struct *t)
```
Comment (evadot) [inline linuxkpi/gplv2/include/linux/interrupt.h:6]: This should be submitted in linuxkpi in base.

**Review by evadot — CHANGES_REQUESTED** [review]

File: linuxkpi/gplv2/include/linux/rcupdate.h (hunk 3 above)
Comment (neelchauhan) [reply linuxkpi/gplv2/include/linux/rcupdate.h:6]: https://reviews.freebsd.org/D30025

File: linuxkpi/gplv2/include/linux/interrupt.h (hunk 4 above)
Comment (neelchauhan) [reply linuxkpi/gplv2/include/linux/interrupt.h:6]: https://reviews.freebsd.org/D30024

**Comment by neelchauhan** [comment]
I have new Phabricator reviews for the functions requested in base:

 * `tasklet_disable_nosync()`: https://reviews.freebsd.org/D30024
 * `rcu_replace_pointer()`: https://reviews.freebsd.org/D30025

**Comment by evadot** [comment]
I've applied some of the patches in 5.5-wip
You still haven't answered to why commenting the perf IOCTL was needed for you (it wasn't for me).
There was other stuff needed to have the module load (See https://github.com/freebsd/drm-kmod/commit/56b414039614b0d83cdbc5bcb2cac04bb64f50d8).
And while I can load the module and have a working console I have a kernel panic (without a chance to do to the debugger) as soon as I start sway (haven't tested Xorg).
So this will stay as 5.5-wip for now.

**Comment by neelchauhan** [comment]
I get crashes with Xorg also. Fortunately, I was able to get a stack trace.

Core dump; [coredump-i915kms-5.5-wsk.txt](https://github.com/freebsd/drm-kmod/files/6431585/coredump-i915kms-5.5-wsk.txt)

This was with a WhiskeyLake HP Spectre x360 13-ap0053dx.

I can also do more tests, and will lalso look into the crashing on my side.

If you want the full core dump, I can upload it. I have a symmetrical uncapped Gigabit connection that I can host a webserver on (what, you'd rather want me to pay double for Comcast?).

**Comment by neelchauhan** [comment]
Here's a log with debugging symbols: [coredump-i915kms-5.5-wsk-dbg.txt](https://github.com/freebsd/drm-kmod/files/6431721/coredump-i915kms-5.5-wsk-dbg.txt)

**Comment by unrelentingtech** [comment]
Hmmmm. So this `intel_context` thing is new, in 5.4 `eb_destroy` did not call this. `intel_context_put` is `kref_put` which is `refcount_release`. Nothing can really go wrong *inside* all this stuff, so this looks like the refcount thingy is just not initialized properly (`fault virtual address	= 0x33be` huh).

And indeed in `reloc_cache_init` I see a suspicious absence of `cache->ce = NULL` !! Which [is present in Linux 5.5](https://github.com/torvalds/linux/blob/v5.5/drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c#L915) soooo someone mismerged something?

**Comment by neelchauhan** [comment]
I think you're correct. I tried this on my WhiskeyLake laptop and Xorg started!

PR: https://github.com/freebsd/drm-kmod/pull/75

**Comment by evadot** [comment]
Handled in another PR.

**Pull request closed without merging**